### PR TITLE
CMR-9848: When submitting multiple associations, if one fails CMR returns a 400 but still makes the other associations

### PR DESCRIPTION
**Pull request by jmaeng72**
# Overview

### What is the problem?

When some associations in POST request, fail, but some succeed. The response returns a 400 for association endpoints, but we want it to reflect the true state that some associations failed.

### What is the Solution?

If a request with multiple associations returns with some successes and some failures, we will return a 207 MULT-STATUS instead, with each association having its own status of 200 or 400 based on success or failure.

Associations affected:

- Tool
- Service
- Concept/Generic
- Tag
- Variable

New Expected Response Status Behavior:

- 200 OK -- if all associations succeeded
- 207 MULTI-STATUS -- if some associations succeeded and some failed due to user error
- 400 BAD REQUEST -- if all associations failed due to user error

### What areas of the application does this impact?

Search App

# Checklist

- [x] I have updated/added unit and int tests that prove my fix is effective or that my feature works
- [x] New and existing unit and int tests pass locally and remotely
- [ ] clj-kondo has been run locally and all errors corrected
- [x] I have removed unnecessary/dead code and imports in files I have changed
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have cleaned up integration tests by doing one or more of the following:
  - migrated any are2 tests to are3 in files I have changed
  - de-duped, consolidated, removed dead int tests
  - transformed applicable int tests into unit tests
  - refactored to reduce number of system state resets by updating fixtures (use-fixtures :each (ingest/reset-fixture {})) to be :once instead of :each

### CODE REVIEW

**Review by eereiter — APPROVED** [review]

File: search-app/src/cmr/search/api/generic_association.clj [hunk 1]
```diff
@@ -1,40 +1,50 @@
...
+    [cmr.search.services.generic-association-service :as generic-assoc-service]
+    [compojure.core :refer :all]))
 
 (defn- validate-association-content-type
   "Validates that content type sent with a association is JSON."
   [headers]
   (mt/extract-header-mime-type #{mt/json} headers "content-type" true))
 
-(defn- api-response
+(defn api-response
```
Comment (jceaser) [inline search-app/src/cmr/search/api/generic_association.clj:19]: why is this now public?

File: search-app/src/cmr/search/api/generic_association.clj [hunk 2]
```diff
@@ -1,40 +1,50 @@
...
-    200)) 
+  (let [result-count (count results)
+        num-errors (assoc/num-errors-in-assoc-results results)]
+    (if (= 0 result-count)
+      200
+      (if (= num-errors result-count)
+        400
+        (if (> num-errors 0)
+          207
+          200)))))
```
Comment (jceaser) [inline search-app/src/cmr/search/api/generic_association.clj:47]: Maybe something like:

    `(let [result-count (count results)
             num-errors (assoc/num-errors-in-assoc-results results)]
          (cond
              (zero? result-count) 200
              (= num-errors result-count) 400
              (pos? num-errors) 207
               :else 200))
    `

Comment (jceaser) [reply search-app/src/cmr/search/api/generic_association.clj:47]: or maybe just drop the 200 test, it's redundant with else.

File: search-app/src/cmr/search/api/tags.clj [hunk 3]
```diff
@@ -5,7 +5,8 @@
    [clojure.string :as string]
    [cmr.acl.core :as acl]
    [cmr.common-app.api.enabled :as common-enabled]
-   [cmr.common.log :refer (debug info warn error)]
+   [cmr.common.log :refer (debug)]
```
Comment (jceaser) [inline search-app/src/cmr/search/api/tags.clj:8]: do you still need () around debug?

File: search-app/src/cmr/search/api/tags.clj [hunk 4]
```diff
@@ -23,16 +24,38 @@
...
+    (let [result-count (count results)
+          num-errors (assoc/num-errors-in-assoc-results results)]
+      (if (= 0 result-count)
+        200
+        (if (= num-errors result-count)
+          400
+          (if (> num-errors 0)
+            207
+            200)))))
+
```
Comment (jceaser) [inline search-app/src/cmr/search/api/tags.clj:44]: a lot of nested if statements. I don't think you need the first one, your last one assumes 200, so you only really testing for 400 and 207. Also, are these constants somewhere in ring? I wonder if the one (if it exists) for 207 is descriptive.

File: search-app/src/cmr/search/api/tags.clj [hunk 5]
```diff
@@ -23,16 +24,38 @@
...
    (if (has-error? data)
      (tag-api-response 400 data)
      (tag-api-response 200 data)))
   ([status-code data]
-   {:status status-code
-    :body (json/generate-string (util/snake-case-data data))
-    :headers {"Content-Type" mt/json}}))
+   (if (= 207 status-code)
+     {:status status-code
+      :body (json/generate-string (util/snake-case-data (cmr.search.api.association/add-individual-statuses data)))
```
Comment (jceaser) [inline search-app/src/cmr/search/api/tags.clj:54]: cmr.search.api.association, you simplify by adding path to imports?

File: system-int-test/src/cmr/system_int_test/utils/association_util.clj [hunk 6]
```diff
@@ -98,3 +98,12 @@
         :body
         first
         :tool-association)))
+
+(defn add-individual-statuses
```
Comment (jceaser) [inline system-int-test/src/cmr/system_int_test/utils/association_util.clj:102]: explain why your creating atoms

File: system-int-test/src/cmr/system_int_test/utils/service_util.clj [hunk 7]
```diff
@@ -192,19 +191,31 @@
   ([coll-service-associations response]
    (assert-service-association-response-ok? coll-service-associations response true))
   ([coll-service-associations response error?]
-   (let [{:keys [status body errors]} response
-         expected-sas (map #(coll-service-association->expected-service-association % error?)
+   (let [{:keys [status body]} response
+         expected-service-assoc (map #(coll-service-association->expected-service-association % error?)
```
Comment (jceaser) [inline system-int-test/src/cmr/system_int_test/utils/service_util.clj:195]: thank you

File: system-int-test/test/cmr/system_int_test/search/misc/association_test.clj [hunk 8]
```diff
@@ -78,10 +78,20 @@
                                                                          :revision-id coll2-revision-id}])
             tla2-concept-id (some #(when (:tool-association %) (get-in % [:tool-association :concept-id]))
                                   (:body assoc-response2))]
-        (is (= {:status 400, :body [{:errors [(str "User doesn't have update permission on INGEST_MANAGEMENT_ACL for provider of collection [" coll1-concept-id "] to make the association.")], :associated-item {:concept-id coll1-concept-id}} {:tool-association {:concept-id tla1-concept-id, :revision-id 1}, :associated-item {:concept-id coll2-concept-id, :revision-id coll2-revision-id}}]}
+        (is (= {:status 207, :body [{:status 400
```
Comment (jceaser) [inline system-int-test/test/cmr/system_int_test/search/misc/association_test.clj:81]: your version is more readable

**Review by jceaser — CHANGES_REQUESTED** [review]

File: search-app/src/cmr/search/api/generic_association.clj (hunk 1 above)
Comment (jmaeng72) [reply search-app/src/cmr/search/api/generic_association.clj:19]: good catch. I made it public to create a unit test real fast. Need to set it back to private

File: search-app/src/cmr/search/api/generic_association.clj (hunk 2 above)
Comment (jmaeng72) [reply search-app/src/cmr/search/api/generic_association.clj:47]: I will need the 200 default for when there are 0 errors and result count > 0

File: search-app/src/cmr/search/api/tags.clj (hunk 3 above)
Comment (jmaeng72) [reply search-app/src/cmr/search/api/tags.clj:8]: you do. It won't work without the parenthesis

File: search-app/src/cmr/search/api/tags.clj (hunk 5 above)
Comment (jmaeng72) [reply search-app/src/cmr/search/api/tags.clj:54]: oh my how did that get in there

File: system-int-test/src/cmr/system_int_test/utils/association_util.clj (hunk 6 above)
Comment (jmaeng72) [reply system-int-test/src/cmr/system_int_test/utils/association_util.clj:102]: I was hoping maybe someone knew a better way to do this. But essentially I am taking a list of maps and need to add an item into each map and return the changed list of maps. I did not know how to do this without using atoms in clojure

Comment (jceaser) [reply system-int-test/src/cmr/system_int_test/utils/association_util.clj:102]: We talked off line, I would suggest that you take the recommended code with a grain of salt and proceed as you see best. I prefer that you do not introduce an atom, but if the code mentioned does not work or causes a delay then roll it back.

File: search-app/src/cmr/search/api/tags.clj (hunk 5 above)
Comment (jmaeng72) [reply search-app/src/cmr/search/api/tags.clj]: fixed

File: system-int-test/src/cmr/system_int_test/utils/association_util.clj (hunk 6 above)
Comment (jmaeng72) [reply system-int-test/src/cmr/system_int_test/utils/association_util.clj:102]: updated atom funcs to remove atom -- AI for the win this time

**Review by jceaser — APPROVED** [review]

**Pull request merged**